soundcraft/installtool.py

```diff
@@ -0,0 +1,934 @@
+
+        # Generate the file contents in Python so we can install it
+        # from Python in case we do have write permissions.
+        lines = [
+            "# Soundcraft Notepad series mixers with audio routing controlled by USB"
+        ]
+        for product_id in const.PY_LIST_OF_PRODUCT_IDS:
+            lines.append(
+                'ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="05fc", ATTRS{idProduct}=="%04x", TAG+="uaccess"'
+                % product_id
+            )
+
+        self.udev_rules_content = "".join([f"{line}\n" for line in lines])
+        self.udev_rules_dst = get_dirs().udev_rulesdir / f"70-{const.PACKAGE}.rules"
+        self.add_file(
+            StringToFile(
+                self.udev_rules_dst,
+                self.udev_rules_content,
+                comment="udev rules allowing non-root access to the USB device",
+            )
+        )
+
+    def emit_code_for_rule_change(self, skip_if):
+        # udev is supposed to be picking up changed rules "for years"
+        # (relative to 2016), so manually triggering a reload does not
+        # appear to be called for any more.
+        #
+        # SUDO_SCRIPT.add_cmd(
+        #     "udevadm control --reload",
+        #     skip_if=skip_if,
+        #     comment="Make udev take notice of the updated set of udev rules",
+        # )
+
+        # FIXME: In case installtool is running as root, run udevadm directly.
+
+        sh_list_of_product_ids = " ".join(
+            ["%04x" % n for n in const.PY_LIST_OF_PRODUCT_IDS]
+        )
+        SUDO_SCRIPT.add_cmd(
+            f"""\
+sleep 4   # wait until udev should have noticed the new rules
+for product_id in {sh_list_of_product_ids}
+do
+    udevadm trigger --verbose \\
+        --action=add --subsystem-match=usb \\
+        --attr-match=idVendor=05fc --attr-match=idProduct=${{product_id}}
+done""",
+            skip_if=skip_if,
+            comment="Trigger udev rules which run when adding existing mixer devices",
+        )
+
+    def post_install(self):
+        # Populate with the files installed pre installation
+        old_content = {}
+        if self.udev_rules_dst.exists():
+            old_content[self.udev_rules_dst] = self.udev_rules_dst.read_text()
+
+        super(UdevRulesInstallTool, self).post_install()
+
+        # Populate with the files which we (should/will) have installed
+        new_content = {}
+        new_content[self.udev_rules_dst] = self.udev_rules_content
+
+        from pprint import pprint
+
+        print("OLD")
+        pprint(old_content)
+        print("NEW")
+        pprint(new_content)
+
+        self.emit_code_for_rule_change(skip_if=(new_content == old_content))
+
+    def pre_uninstall(self):
+        # FIXME05: Do we even want to uninstall the udev rules if it
+        #          is in /etc/udev/rules.d for a $HOME/.local install?
+        #          The next install will just need sudo access again
+        #          to install it again.
```

Comment:
I think on a true uninstall (as in "this package is going away"), yes.  We should remove the udev rules.

If this hook also runs as part of an upgrade, we'll have to be more clever: We only want to touch the udev rules if the new version's rules are different than what's on the system already.